### PR TITLE
dropbear: expose optional scp support

### LINKED ISSUE #135849: `dropbear`: add option to build `scp` utility
`dropbear` is already packaged, but I needed the following to be able to correctly build the `scp` utility:
```nix
with pkgs;
   (dropbear.overrideAttrs (old: {
      preConfigure = old.preConfigure + ''
        makeFlagsArray=( "$makeFlags" "PROGRAMS=dbclient scp" )
        makeFlags=""
      '';
      CFLAGS = old.CFLAGS + " -DDROPBEAR_PATH_SSH_PROGRAM=\\\"${placeholder "out"}/bin/dbclient\\\"";
    }))
```
I would propose adding an option `withSCP ? false`, which adds `scp` to the default `PROGRAMS` list `dropbear dbclient dropbearkey dropbearconvert`, and also modifies `CFLAGS` to include the fix of `DROPBEAR_PATH_SSH_PROGRAM`. It might be a good idea to also mark `dropbear` as `lowPrio` to prevent accidental conflicts with `openssh`.

### PR DESCRIPTION
**Pull request by fogti**
###### Motivation for this change

Fixes #135849.
I have a piece of old hardware with such an old SSH installation that it is only supported by `dropbear`, and I want to automatically copy files onto it, thus I need the `scp` utility. This was already packaged on gentoo, and was missing on NixOS.

cc @abbradar

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`) (tested only `dbclient` and `scp`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/networking/dropbear/default.nix [hunk 1]
```diff
@@ -12,24 +31,35 @@ stdenv.mkDerivation rec {
...
-    # Allow sessions to inherit the PATH from the parent dropbear.
-    # Otherwise they only get the usual /bin:/usr/bin kind of PATH
-    ./pass-path.patch
-  ];
-
-  buildInputs = [ zlib ] ++ lib.optionals enableStatic [ glibc.static zlib.static ];
+  postInstall = lib.optionalString enableSCP ''
+    pushd $out/bin
+    ln -s scp dbscp
+    popd
```
Comment (kvtb) [inline pkgs/tools/networking/dropbear/default.nix:61]:
```suggestion
    ln -s scp $out/bin/dbscp
```

File: pkgs/tools/networking/dropbear/default.nix [hunk 2]
```diff
@@ -1,8 +1,27 @@
-{ stdenv, lib, fetchurl, glibc, zlib
+{ stdenv
+, lib
+, fetchurl
+, glibc
+, zlib
 , enableStatic ? stdenv.hostPlatform.isStatic
```
Comment (zseri) [inline pkgs/tools/networking/dropbear/default.nix:6]: wouldn't it be a better idea to put this support into [`pkgs/top-level/static.nix`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/static.nix)?

File: pkgs/tools/networking/dropbear/default.nix [hunk 3]
```diff
@@ -1,8 +1,27 @@
...
+, enableSCP ? false
 , sftpPath ? "/run/current-system/sw/libexec/sftp-server"
 }:
 
+let
+  inherit (builtins) attrNames concatStringsSep map;
+
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/dropbear/default.nix:13]:
```suggestion
```

File: pkgs/tools/networking/dropbear/default.nix [hunk 4]
```diff
@@ -12,24 +31,33 @@ stdenv.mkDerivation rec {
...
+    (map (name: "-D${name}=\\\"${dflags.${name}}\\\""))
+    (concatStringsSep " ")
+  ];
 
   # https://www.gnu.org/software/make/manual/html_node/Libraries_002fSearch.html
   preConfigure = ''
-    makeFlags=VPATH=`cat $NIX_CC/nix-support/orig-libc`/lib
+    makeFlagsArray=(
+      VPATH=`cat $NIX_CC/nix-support/orig-libc`/lib
+      PROGRAMS="${concatStringsSep " " programs}"
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/dropbear/default.nix:54]:
```suggestion
      PROGRAMS="${lib.concatStringsSep " " [ "dropbear" "dbclient" "dropbearkey" "dropbearconvert" ] ++ lib.optionals enableSCP ["scp"];
}"
```

File: pkgs/tools/networking/dropbear/default.nix [hunk 5]
```diff
@@ -12,24 +31,33 @@ stdenv.mkDerivation rec {
...
 
+  dontDisableStatic = enableStatic;
   configureFlags = lib.optional enableStatic "LDFLAGS=-static";
 
-  CFLAGS = "-DSFTPSERVER_PATH=\\\"${sftpPath}\\\"";
+  buildInputs = [ zlib ] ++ lib.optionals enableStatic [ glibc.static zlib.static ];
+
+  CFLAGS = lib.pipe (attrNames dflags) [
+    (map (name: "-D${name}=\\\"${dflags.${name}}\\\""))
+    (concatStringsSep " ")
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/dropbear/default.nix:47]:
```suggestion
    (lib.concatStringsSep " ")
```

File: pkgs/tools/networking/dropbear/default.nix [hunk 6]
```diff
@@ -12,24 +31,33 @@ stdenv.mkDerivation rec {
...
+  ];
 
+  dontDisableStatic = enableStatic;
   configureFlags = lib.optional enableStatic "LDFLAGS=-static";
 
-  CFLAGS = "-DSFTPSERVER_PATH=\\\"${sftpPath}\\\"";
+  buildInputs = [ zlib ] ++ lib.optionals enableStatic [ glibc.static zlib.static ];
+
+  CFLAGS = lib.pipe (attrNames dflags) [
+    (map (name: "-D${name}=\\\"${dflags.${name}}\\\""))
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/dropbear/default.nix:46]:
```suggestion
    (builtins.map (name: "-D${name}=\\\"${dflags.${name}}\\\""))
```

File: pkgs/tools/networking/dropbear/default.nix [hunk 7]
```diff
@@ -12,24 +31,33 @@ stdenv.mkDerivation rec {
...
+    ./pass-path.patch
+  ];
 
+  dontDisableStatic = enableStatic;
   configureFlags = lib.optional enableStatic "LDFLAGS=-static";
 
-  CFLAGS = "-DSFTPSERVER_PATH=\\\"${sftpPath}\\\"";
+  buildInputs = [ zlib ] ++ lib.optionals enableStatic [ glibc.static zlib.static ];
+
+  CFLAGS = lib.pipe (attrNames dflags) [
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/dropbear/default.nix:45]:
```suggestion
  CFLAGS = lib.pipe (lib.attrNames dflags) [
```

File: pkgs/tools/networking/dropbear/default.nix [hunk 8]
```diff
@@ -1,8 +1,27 @@
...
 , enableStatic ? stdenv.hostPlatform.isStatic
+, enableSCP ? false
 , sftpPath ? "/run/current-system/sw/libexec/sftp-server"
 }:
 
+let
+  inherit (builtins) attrNames concatStringsSep map;
+
+  programs = [ "dropbear" "dbclient" "dropbearkey" "dropbearconvert" ] ++ lib.optionals enableSCP ["scp"];
+
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/dropbear/default.nix:15]:
```suggestion
```

File: pkgs/tools/networking/dropbear/default.nix [hunk 9]
```diff
@@ -12,24 +27,33 @@ stdenv.mkDerivation rec {
...
+    (builtins.map (name: "-D${name}=\\\"${dflags.${name}}\\\""))
+    (lib.concatStringsSep " ")
+  ];
 
   # https://www.gnu.org/software/make/manual/html_node/Libraries_002fSearch.html
   preConfigure = ''
-    makeFlags=VPATH=`cat $NIX_CC/nix-support/orig-libc`/lib
+    makeFlagsArray=(
+      VPATH=`cat $NIX_CC/nix-support/orig-libc`/lib
+      PROGRAMS="${lib.concatStringsSep " " [ "dropbear" "dbclient" "dropbearkey" "dropbearconvert" ] ++ lib.optionals enableSCP ["scp"]}"
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/dropbear/default.nix:50]:
```suggestion
      PROGRAMS="${lib.concatStringsSep " " ([ "dropbear" "dbclient" "dropbearkey" "dropbearconvert" ] ++ lib.optionals enableSCP ["scp"])}"
```

I think we are missing those.

**Review by SuperSandro2000 — COMMENTED** [review]
I wouldn't mind if we moved buildInputs back to where it was to reduce the diff but not that important.

**Comment by zseri** [comment]
ehm @abbradar could you please take a look at this?

**Review by fpletz — APPROVED** [review]

**Pull request merged**